rbac: add user-facing docs (#49472)
sourcegraph/sourcegraph: doc/admin/access_control/index.md

> # Access control

Author: That's a good point. I'll leave it alone for now as to not bloat this PR, but let's definitely keep that in mind as we continue to work in this area of the docs!

<aside class="beta">
<p>
<span class="badge badge-beta">Beta</span> <strong>This feature is currently in beta.</strong>
</p>
</aside>

> NOTE: This page refers to in-product permissions, which determine who can, for example, create a batch change, or who is a site admin. This is *not* the same as [repository permissions](../permissions/index.md), which enforces the same repository access on Sourcegraph as your code host.

> > NOTE: This functionality was added in Sourcegraph 5.0.

Author: Ah good to know, we've not been using this in the Batch Changes docs and, as you might have observed, I got all my patterns from the Batch Changes docs. 😂 Will clean that up.

Sourcegraph uses [Role-Based Access Control (RBAC)](https://en.wikipedia.org/wiki/Role-based_access_control) to enable fine-grained control over different features and abilities of Sourcegraph, without having to modify permissions for each user individually. Currently, the scope of permissions control is limited to [Batch Changes](batch_changes.md) functionality, but it will be expanded to other areas in the future.

## Managing roles and permissions

<!-- TODO: Screenshots -->

> Site admins can control which features each type of user has access to by creating custom roles and assigning permissions to them. You can see all available roles and create new ones under _Site admin_ > _Users & auth_ > _Roles_.

Author: It's things like this that really made me wish we just could properly componentize things in our docs. 😢 Thanks for taking the time to do the research!

> Site admins can control which features each type of user has access to by creating custom roles and assigning permissions to them. You can see all available roles and create new ones under _Site admin_ > _Users & auth_ > _Roles_.

Reviewer: Yes, I was thinking about the same things. It's hard to keep the conventions when it's basically free for all format in markdown.

### System roles

> Every Sourcegraph instance ships with two system roles:

Author: Haha here's another fun one, [builtin](https://sourcegraph.sourcegraph.com/search?q=context:sourcegraph+file:doc/(.*/)*.*%5C.md+builtin&patternType=regexp&sm=1&groupBy=repo) or [built-in](https://sourcegraph.sourcegraph.com/search?q=context:sourcegraph+file:doc/(.*/)*.*%5C.md+built-in&patternType=regexp&sm=1&groupBy=repo)? 😄

- **Site Administrator**: This role is granted to any user who is promoted to site admin. It always has all features and permissions of Sourcegraph granted to it.
- **User:** This role is granted to every user of the Sourcegraph instance and cannot be unassigned. By default, it has all features and permissions of Sourcegraph granted to it, but this can be modified.

### Creating a new role and assigning it permissions

To create a new role, click the **+ Create role** button. Give the role a unique, descriptive name, then select which permissions to associate with it using the checkboxes. Then click **Create**.

### Editing permissions for an existing role

> With the exception of the **Site Administrator** role, all role permissions can be modified. To edit the permissions granted to a role, click the role to expand it, then select the new set of permissions you want to grant to it. Then click **Update** to save your changes.

Reviewer: ```suggestion

> NOTE: **Site Administrator** role cannot be modified.

To edit the permissions granted to a role, click the role to expand it, then select the new set of permissions you want to grant to it. Then click **Update** to save your changes.
```

You can read about the specific permission types available for each RBAC-enabled product area below:

> - [Batch Changes](./batch_changes.md)

Reviewer: Here the link is relative `./batch_changes.md`, on line 13 it does not start with `./`. Probably makes sense to use the same link in both places, would be easier to search & replace in the future if we move the page.

Author: Ha you're right, good eyes! 👁️

> NOTE: While Batch Changes is the only RBAC-enabled product area today, we will be working on migrating other product areas in future releases of Sourcegraph. Please reach out to our [support team](mailto:support@sourcegraph.com) if you have further questions. 

### Deleting a role

> With the exception of system roles, all roles can be deleted. To delete a role, click the **Delete** button on it. You will be prompted to confirm your choice. Once deleted, any user previously assigned that role will lose any permissions associated with it that are not also granted by their other roles.

Reviewer: ```suggestion

> NOTE: Builtin system roles cannot be deleted.

To delete a role, click the **Delete** button on it. You will be prompted to confirm your choice. Once deleted, all users previously assigned that role will lose all permissions associated with the deleted role. Be aware though that same permissions might also granted by their other roles.
```

## Managing user roles

> Site admins can manage which roles are assigned to which users from _Site admin_ > _Users & auth_ > _Users_. To view or edit a user's roles, click the triple dots to open the context menu for that user, then click **Manage roles**. This will open a modal where you can see the user's current roles, assign new ones, or unassign current ones. You can type in the input field to search roles by name. Click **Update** to save any changes, or **Cancel** to discard. Note that system roles cannot be revoked or assigned via this modal.

Reviewer: ```suggestion

> NOTE: Builtin system roles cannot be assigned this way.

Site admins can manage role assignments from **Site admin > Users & auth > Users**. To view or edit a user's roles, click the triple dots to open the context menu for that user, then click **Manage roles**. This will open a modal dialog where you can see the user's current roles, assign new ones, or unassign current ones. You can type in the input field to search roles by name. Click **Update** to save any changes, or **Cancel** to discard.
```

> To assign the **Site Administrator** system role to a user, open the same context menu from the triple dots, then click **Promote to site admin**. To unassign the **Site Administrator** role, open the same context menu from the triple dots, then click **Revoke site admin**.

Reviewer: ```suggestion
To assign the **Site Administrator** system role to a user, open the same context menu from the triple dots, then click **Promote to site admin**. To unassign the **Site Administrator** role, open the same context menu from the triple dots, then click **Revoke site admin**.

The **User** system role is automatically assigned to all users and cannot be revoked.
```
